NOnion/Utility/Base32Util.fs

```diff
@@ -0,0 +1,171 @@
+                    '='
+                    '='
+                    '='
+                    '='
+                    '='
+                ] // 01======
+            | 2 ->
+                [
+                    alphabet.[a0]
+                    alphabet.[a1]
+                    alphabet.[a2]
+                    alphabet.[a3]
+                    '='
+                    '='
+                    '='
+                    '='
+                ] // 0123====
+            | 3 ->
+                [
+                    alphabet.[a0]
+                    alphabet.[a1]
+                    alphabet.[a2]
+                    alphabet.[a3]
+                    alphabet.[a4]
+                    '='
+                    '='
+                    '='
+                ] // 01234===
+            | 4 ->
+                [
+                    alphabet.[a0]
+                    alphabet.[a1]
+                    alphabet.[a2]
+                    alphabet.[a3]
+                    alphabet.[a4]
+                    alphabet.[a5]
+                    alphabet.[a6]
+                    '='
+                ] // 0123456=
+            | _ ->
+                [
+                    alphabet.[a0]
+                    alphabet.[a1]
+                    alphabet.[a2]
+                    alphabet.[a3]
+                    alphabet.[a4]
+                    alphabet.[a5]
+                    alphabet.[a6]
+                    alphabet.[a7]
+                ] // 01234567
+
+        let rec parse result input =
+            match input with
+            | x0 :: x1 :: x2 :: x3 :: x4 :: tail ->
+                parse (result @ quintupletToList 5 (x0, x1, x2, x3, x4)) tail
+            | x0 :: x1 :: x2 :: [ x3 ] ->
+                result @ quintupletToList 4 (x0, x1, x2, x3, 0uy)
+            | x0 :: x1 :: [ x2 ] ->
+                result @ quintupletToList 3 (x0, x1, x2, 0uy, 0uy)
+            | x0 :: [ x1 ] ->
+                result @ quintupletToList 2 (x0, x1, 0uy, 0uy, 0uy)
+            | [ x0 ] -> result @ quintupletToList 1 (x0, 0uy, 0uy, 0uy, 0uy)
+            | [] -> result
+
+        (text: string)
+        |> System.Text.Encoding.UTF8.GetBytes
+        |> Array.toList
+        |> parse []
+        |> List.toArray
+        |> System.String.Concat
+
+    /// Decodes a Base32 string to a UTF8 string
+    let DecodeBase32 text =
+        if (text: string).Length % 8 <> 0 then
+            [||]
+        else
+            // RFC 4648: The Base 32 alphabet
+            let alphabet =
+                [
+                    for c in "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567=" -> c
+                ]
+                |> List.mapi(fun i a -> a, i)
+                |> Map.ofList
+
+            // RFC 4648: The "Extended Hex" Base 32 alphabet
+            // let A = [for c in "0123456789ABCDEFGHIJKLMNOPQRSTUV=" -> c]
+            //         |> List.mapi (fun i a -> a, i)
+            //         |> Map.ofList
+
+            let (.@) (m: Map<char, int>) key =
```

Comment:
@aarani what is this .@ thing?

Reply:
This code is copy-paste from somewhere, I have no clue what it does.

Reply:
I mean I know what it does but no clue how it does it

Reply:
then put the URL of that somewhere

Reply:
> I know what it does

explain it in a comment then

Reply:
> then put the URL of that somewhere

It's there in a commit

Reply:
> > I know what it does
> 
> explain it in a comment then

Seriously? EncodeToBase32 is not enough? what should I add? what is base32?

Reply:
I'm not talking about the outer function, I'm talking about this inner function

Reply:
When I said I know what it does I mean I know what the entire function does, not this inner function.